src/include/sof/audio/buffer.h

```diff
@@ -343,6 +345,43 @@ static inline void buffer_copy(struct comp_buffer *source,
 	}
 }
 
+static inline int buffer_set_params(struct comp_buffer *buffer,
+				    struct sof_ipc_stream_params *params)
+{
+	int i;
+
+	if (!params) {
+		trace_buffer_error("buffer_set_params() error: !params");
+		return -EINVAL;
+	}
+
+	buffer->frame_fmt = params->frame_fmt;
+	buffer->buffer_fmt = params->buffer_fmt;
+	buffer->rate = params->rate;
+	buffer->channels = params->channels;
+	for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
+		buffer->chmap[i] = params->chmap[i];
+
+	return 0;
+}
+
+static inline int buffer_set_hw_params(struct comp_buffer *buffer,
+				       struct sof_ipc_stream_params *params)
+{
+	int ret = 0;
+
+	tracev_buffer("buffer_set_hw_params(): buffer->id: %d", buffer->id);
+
+	ret = buffer_set_params(buffer, params);
+	if (ret < 0) {
+		trace_buffer_error("buffer_set_hw_params() error: buffer_set_params() failed");
+	}
+
+	buffer->hw_params_configured = true;
```

Comment:
Sorry, this should be set here ```buffer_set_params()```, to be honest is there any point in having ```buffer_set_hw_params()``` and ```buffer_set_params()``` as one is just a thin wrapper for the other ?

Reply:
@bkokoszx ping ?

Reply:
I've added buffer->hw_params_configured in order not to set hardware params in buffer twice.
```
pipe1									                   pipe3
[HOST] -> [BUF3] -> [MUX] ->[BUF1] -> [COMP] -> [BUF0] -> [DAI0]
                      |       
[HOST] -> [BUF2] ---> +
pipe2
```

If we call params() on pipe1, we will set hw parameters on pipe3 and pipe1. Then if we call params() on pipe2, we will not set hw parameters on pipe3, because ```hw_params_configured``` has been set in previous step.

I've added ```buffer_set_hw_params()``` wrapper in order not to set ```buffer->hw_params_configured``` directly from pipeline.c file, but of course I can remove that wrapper and make:
``` diff
diff --git a/src/audio/pipeline.c b/src/audio/pipeline.c
index 40fb95cd..ed2f7430 100644
--- a/src/audio/pipeline.c
+++ b/src/audio/pipeline.c
@@ -296,8 +296,10 @@ static int pipeline_comp_hw_params(struct comp_dev *current,
        }

        /* set buffer parameters */
-       if (calling_buf && !calling_buf->hw_params_configured)
-               buffer_set_hw_params(calling_buf, &ppl_data->params->params);
+       if (calling_buf && !calling_buf->hw_params_configured) {
+               buffer_set_params(calling_buf, &ppl_data->params->params);
+               calling_buf->hw_params_configured = true;
+       }

        return ret;
 }

```

Reply:
ok, it would be better if 
```
buffer_set_params(buf)
{
    if (buf->hw_params_configured)
        return;

     buf->rate = X;
     buf->format = Y;
     buf->hw_params_configured = true;
    return;
}
```
This means we dont have to put this check into pipeline.c as it's contained in the buffer logic.

Reply:
@lgirdwood 
We cannot do that, because ```buffer_set_params()``` (assuming that we do not have ```buffer_set_hw_params()```) is invoked two times: 
- first in ```pipeline_comp_hw_params()```, where we set hardware stream parameters (```buffer->hw_params_configured``` will be set to true)
- second in ```pipeline_comp_params()```, where we set pcm parameters. Here, buffers already have ```hw_params_configured``` flag set and ```buffer_set_params()``` would stop at:
```
if (buf->hw_params_configured)
        return;
```

Reply:
ok, so it sounds like one API call only sets the params if they are not already configured and the other API forces the new params ?  if so we can have 
```
buffer_set_params();
buffer_set_params_forced();
```
or we could pass a bool flag as to whether to force update the params i.e.
```
buffer_set_params(..., bool force_update);
```
This would flow better and make the updates easier to follow.